File: Frequency Distribution/Kotlin/FrequencyDistributionMap.kt
import java.util.ArrayList
import java.util.Arrays
import kotlin.math.log10
import kotlin.math.roundToInt

class FrequencyDistributionMap(private val data: IntArray) {
    val size: Int = data.size
    val min: Int
    val max: Int
    val range: Int
    val classAmount: Long
    val classLength: Int
    private val _intervalBottom: MutableList<Int> = ArrayList()
    private val _intervalTop: MutableList<Int> = ArrayList()
    private val frequency: MutableList<Int> = ArrayList()
    private val classMark: MutableList<Double> = ArrayList()
    private val trueLowerLimit: MutableList<Double> = ArrayList()
    private val trueUpperLimit: MutableList<Double> = ArrayList()
    private val cumulativeFrequencyLessThanType: MutableList<Int> = ArrayList()
    private val cumulativeFrequencyGreaterThanType: MutableList<Int> = ArrayList()
    fun getInterval(index: Int, splitter: String): String {
        return _intervalBottom[index].toString() + splitter + _intervalTop[index]
    }

    fun getIntervalBottom(index: Int): Int {
        return _intervalBottom[index]
    }

    fun getIntervalTop(index: Int): Int {
        return _intervalTop[index]
    }

    fun getFrequency(index: Int): Int {
        return frequency[index]
    }

    fun getClassMark(index: Int): Double {
        return classMark[index]
    }

    fun getClassBoundaries(index: Int, splitter: String): String {
        return trueLowerLimit[index].toString() + splitter + trueUpperLimit[index]
    }

    fun getTrueLowerLimit(index: Int): Double {
        return trueLowerLimit[index]
    }

    fun getTrueUpperLimit(index: Int): Double {
        return trueUpperLimit[index]
    }

    fun getCumulativeFrequencyLessThanType(index: Int): Int {
        return cumulativeFrequencyLessThanType[index]
    }

    fun getCumulativeFrequencyGreaterThanType(index: Int): Int {
        return cumulativeFrequencyGreaterThanType[index]
    }

    val row: Int
        get() = _intervalBottom.size
    val mean: Double
        get() {
            var total = 0.0
            for (i in 0 until classAmount) {
                total += frequency[i.toInt()] * classMark[i.toInt()]
            }
            return total / size
        }

    private fun setupInterval() {
        for (i in 0 until classAmount) {
            addInterval((min + (classLength + 1) * i).toInt(), (min + classLength + (classLength + 1) * i).toInt())
        }
    }

    private fun setupFrequency() {
        for (i in 0 until size) {
            for (j in 0 until classAmount) {
                if (data[i] >= _intervalBottom[j.toInt()] && data[i] <= _intervalTop[j.toInt()]) {
                    var count = frequency[j.toInt()]
                    count++
                    frequency[j.toInt()] = count
                }
            }
        }
    }

    private fun setupLessAndGreater() {
        var countLess = 0
        var countGreater = size
        for (i in 0 until classAmount) {
            countLess += frequency[i.toInt()]
            cumulativeFrequencyLessThanType.add(countLess)
            countGreater -= frequency[i.toInt()]
            cumulativeFrequencyGreaterThanType.add(countGreater)
        }
    }

    private fun addInterval(intervalBottom: Int, intervalTop: Int) {
        _intervalBottom.add(intervalBottom)
        _intervalTop.add(intervalTop)
        frequency.add(0)
        classMark.add(intervalBottom.toDouble() + (intervalTop.toDouble() - intervalBottom.toDouble()) / 2)
        trueLowerLimit.add(intervalBottom.toDouble() - 0.5)
        trueUpperLimit.add(intervalTop.toDouble() + 0.5)
    }

    init {
        val stat = Arrays.stream(data).summaryStatistics()
        min = stat.min
        max = stat.max
        range = max - min
        classAmount = (1 + 3.3 * log10(data.size.toDouble())).roundToInt().toLong()
        classLength = range / classAmount.toFloat().roundToInt()
        setupInterval()
        setupFrequency()
        setupLessAndGreater()
    }
}

File: Frequency Distribution/c#/FrequencyDistributionMap.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

class FrequencyDistributionMap
{
    private int[] data;
    private int size;
    private int min;
    private int max;
    private int range;
    private long classAmount;
    private int classLength;

    private IList<int> _intervalBottom = new List<int>();
    private IList<int> _intervalTop = new List<int>();
    private IList<int> frequency = new List<int>();
    private IList<double> classMark = new List<double>();
    private IList<double> trueLowerLimit = new List<double>();
    private IList<double> trueUpperLimit = new List<double>();
    private IList<int> cumulativeFrequencyLessThanType = new List<int>();
    private IList<int> cumulativeFrequencyGreaterThanType = new List<int>();

    public FrequencyDistributionMap(int[] data)
    {
        this.data = data;
        size = data.Length;
        min = data.Min();
        max = data.Max();
        range = max - min;
        classAmount = (long)Math.Round(1 + ((3.3) * Math.Log10(data.Length)), MidpointRounding.AwayFromZero);
        classLength = range / (int)Math.Round(Convert.ToDouble(classAmount));
        setupInterval();
        setupFrequency();
        setupLessAndGreater();
    }

    public virtual string getInterval(int index, string splitter)
    {
        return _intervalBottom[index] + splitter + _intervalTop[index];
    }

    public virtual int getIntervalBottom(int index)
    {
        return _intervalBottom[index];
    }

    public virtual int getIntervalTop(int index)
    {
        return _intervalTop[index];
    }

    public virtual int getFrequency(int index)
    {
        return frequency[index];
    }

    public virtual double getClassMark(int index)
    {
        return classMark[index];
    }

    public virtual string getClassBoundaries(int index, string splitter)
    {
        return trueLowerLimit[index] + splitter + trueUpperLimit[index];
    }

    public virtual double getTrueLowerLimit(int index)
    {
        return trueLowerLimit[index];
    }

    public virtual double getTrueUpperLimit(int index)
    {
        return trueUpperLimit[index];
    }

    public virtual int getCumulativeFrequencyLessThanType(int index)
    {
        return cumulativeFrequencyLessThanType[index];
    }

    public virtual int getCumulativeFrequencyGreaterThanType(int index)
    {
        return cumulativeFrequencyGreaterThanType[index];
    }

    public virtual int Row
    {
        get
        {
            return _intervalBottom.Count;
        }
    }

    public virtual int Size
    {
        get
        {
            return size;
        }
    }

    public virtual int Min
    {
        get
        {
            return min;
        }
    }

    public virtual int Max
    {
        get
        {
            return max;
        }
    }

    public virtual double? Mean
    {
        get
        {
            double? total = 0.0;
            for (int i = 0; i < classAmount; i++)
            {
                total = total.Value + (frequency[i] * classMark[i]);
            }
            return total.Value / Size;
        }
    }

    public virtual int Range
    {
        get
        {
            return range;
        }
    }

    public virtual long? ClassAmount
    {
        get
        {
            return classAmount;
        }
    }

    public virtual int ClassLength
    {
        get
        {
            return classLength;
        }
    }

    private void setupInterval()
    {
        for (int i = 0; i < classAmount; i++)
        {
            addInterval((int)(Min + ((ClassLength + 1) * i)), (int)((Min + ClassLength) + ((ClassLength + 1) * i)));
        }
    }

    private void setupFrequency()
    {
        for (int i = 0; i < Size; i++)
        {
            for (int j = 0; j < classAmount; j++)
            {
                if (data[i] >= _intervalBottom[j] && data[i] <= _intervalTop[j])
                {
                    int count = frequency[j];
                    count++;
                    frequency[j] = count;
                }
            }
        }
    }

    private void setupLessAndGreater()
    {
        int countLess = 0;
        int countGreater = Size;
        for (int i = 0; i < classAmount; i++)
        {
            countLess += frequency[i];
            cumulativeFrequencyLessThanType.Add(countLess);
            countGreater -= frequency[i];
            cumulativeFrequencyGreaterThanType.Add(countGreater);
        }
    }

    private void addInterval(int intervalBottom, int intervalTop)
    {
        _intervalBottom.Add(intervalBottom);
        _intervalTop.Add(intervalTop);
        frequency.Add(0);
        classMark.Add((double)intervalBottom + (((double)intervalTop - (double)intervalBottom) / 2));
        trueLowerLimit.Add((double)intervalBottom - 0.5);
        trueUpperLimit.Add((double)intervalTop + 0.5);
    }
}
